PEP 659: Update to describe inline caches (#2462)
python/peps: pep-0659.rst

Performance
-----------

The speedup from specialization is hard to determine, as many specializations
depend on other optimizations. Speedups seem to be in the range 10% - 60%.

* Most of the speedup comes directly from specialization. The largest
  contributors are speedups to attribute lookup, global variables, and calls.
* A small, but useful, fraction is from from improved dispatch such as
  super-instructions and other optimizations enabled by quickening.

Implementation
==============

Overview
--------

> Any instruction that would benefit from specialization will be replaced by an

Reviewer: Any reason why you don't discuss the warmup period anymore? Perhaps just because we're moving towards quickening everything now?

Author: Yes, quickening everything, means that "warm up" is handled by the adaptive instructions.

adaptive version during quickening. For example,
the ``LOAD_ATTR`` instruction would be replaced with ``LOAD_ATTR_ADAPTIVE``.

Each adaptive instruction maintains a counter, and periodically attempts

> to specialize itself.

Reviewer: There is an almost-identical sentence two paragraphs down. Maybe remove one of them?

--------------

Most families of specialized instructions will require more information than
can fit in an 8-bit operand. To do this, a number of 16 bit entries immediately
following the instruction are used to store this data. This is a form of inline
cache, an "inline data cache". Unspecialized, or adaptive, instructions will
use the first entry of this cache as a counter, and simply skip over the others.

Example families of instructions
--------------------------------

LOAD_ATTR
'''''''''

The ``LOAD_ATTR`` loads the named attribute of the object on top of the stack,
then replaces the object on top of the stack with the attribute.

This is an obvious candidate for specialization. Attributes might belong to
a normal instance, a class, a module, or one of many other special cases.

``LOAD_ATTR`` would initially be quickened to ``LOAD_ATTR_ADAPTIVE`` which

> would track how often it is executed, and call the ``load_attr_optimize``

Reviewer: ```suggestion
would track how often it is executed, and call the ``_Py_Specialize_LoadAttr``
```

An obvious concern with any scheme that performs any sort of caching is
"how much more memory does it use?".
The short answer is "not that much".

> Memory use in 3.10
> ''''''''''''''''''
> 
> CPython 3.10 used 2 bytes per instruction, until the execution count
> reached ~2000 when another byte per instruction and 32 bytes per
> instruction with a cache (``LOAD_GLOBAL`` and ``LOAD_ATTR``).
> 
> Comparing memory use to 3.10
> ''''''''''''''''''''''''''''
> 
> CPython 3.10 used 2 bytes per instruction, until the execution count
> reached ~2000 when it allocates another byte per instruction and
> 32 bytes per instruction with a cache (``LOAD_GLOBAL`` and ``LOAD_ATTR``).
> 

Reviewer: This paragraph is repeated (and the first copy looks incomplete).

```suggestion
Comparing memory use to 3.10
''''''''''''''''''''''''''''

CPython 3.10 used 2 bytes per instruction, until the execution count
reached ~2000 when it allocates another byte per instruction and
32 bytes per instruction with a cache (``LOAD_GLOBAL`` and ``LOAD_ATTR``).

```

The following table shows the additional bytes per instruction to support the
3.10 opcache or the proposed adaptive interpreter, on a 64 bit machine.

================   ==========  ==========  ======
 Version           3.10 cold    3.10 hot    3.11
 Specialised           0%        ~15%       ~25%
----------------   ----------  ----------  ------
 code                 2           2          2
 opcache_map          0           1          0
 opcache/data         0          4.8         4
----------------   ----------  ----------  ------
 Total                2          7.8         6
================   ==========  ==========  ======

``3.10 cold`` is before the code has reached the ~2000 limit.
``3.10 hot`` shows the cache use once the threshold is reached.

The relative memory use depends on how much code is "hot" enough to trigger
creation of the cache in 3.10. The break even point, where the memory used
by 3.10 is the same as for 3.11 is ~70%.

It is also worth noting that the actual bytecode is only part of a code
object. Code objects also include names, constants and quite a lot of

> debugging information.

Reviewer: Maybe touch on the line/column-table issue when adding additional cache entries? Could be worded as "the format of these (cold) tables may need to be updated to avoid additional memory use" or something.

Author: I'm not sure that is relevant. We would probably want to revisit how we store the debug information even if we kept the original bytecode format.

> In summary, for most applications where many of the functions are relatively
> relatively unused, 3.11 will consume more memory than 3.10, but not by much.

Reviewer: ```suggestion
In summary, for most applications where many of the functions are relatively
unused, 3.11 will consume more memory than 3.10, but not by much.
```

Rejected Ideas
==============

By implementing a specializing adaptive interpreter with inline data caches,
we are implicitly rejecting many alternative ways to optimize CPython.
However, it is worth emphasizing that some ideas, such as just-in-time
compilation, have not been rejected, merely deferred.

> Storing data caches before the bytecode.

Reviewer: Do you think it's worth enumerating a few reasons why we decided not to go this route?

Author: Added a brief comment.
